packages/bp/src/core/nlu/nlu-inference-service.ts

```diff
@@ -0,0 +1,161 @@
+import { IO, Logger } from 'botpress/sdk'
+import { ConfigProvider } from 'core/config'
+import { WellKnownFlags } from 'core/dialog'
+import { EventEngine } from 'core/events'
+import { TYPES } from 'core/types'
+import { inject, injectable, postConstruct } from 'inversify'
+import { AppLifecycle, AppLifecycleEvents } from 'lifecycle'
+import _ from 'lodash'
+import yn from 'yn'
+import legacyElection from './election/legacy-election'
+import naturalElection from './election/natural-election'
+import pickSpellChecked from './election/spellcheck-handler'
+import { NLUClientProvider } from './nlu-client'
+import { Predictor } from './predictor'
+
+const EVENTS_TO_IGNORE = ['session_reference', 'session_reset', 'bp_dialog_timeout', 'visit', 'say_something', '']
+const PREDICT_MW = 'nlu-predict.incoming'
+
+/**
+ * This service takes care of the nlu inferences/predictions
+ *
+ * It belongs in the runtime and only in the runtime.
+ * It is not responsible for CRUD on intent and entity files.
+ * It is not responsible for trainings and the models state-machine.
+ * It expects bot config to have field nluModels.
+ * It expects botpress config to have field nlu.
+ * That's it.
+ */
+@injectable()
+export class NLUInferenceService {
+  private legacyElection!: boolean
+
+  private predictors: { [botId: string]: Predictor } = {}
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider,
+    @inject(TYPES.Logger) private logger: Logger,
+    @inject(TYPES.NLUClientProvider) private nluClientProvider: NLUClientProvider
+  ) {}
+
+  @postConstruct()
+  public async initialize() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+    await this.nluClientProvider.initialize()
+    const { nlu: nluConfig } = await this.configProvider.getBotpressConfig()
+
+    const { legacyElection } = nluConfig
+    this.legacyElection = legacyElection ?? false
+
+    this.eventEngine.register({
+      name: PREDICT_MW,
+      description:
+        'Process natural language in the form of text. Structured data with an action and parameters for that action is injected in the incoming message event.',
+      order: 100,
+      direction: 'incoming',
+      handler: this.handleIncomingEvent.bind(this)
+    })
+  }
+
+  public async teardown() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+    this.eventEngine.removeMiddleware(PREDICT_MW)
+  }
+
+  public async mountBot(botId: string) {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
```

Comment:
I don't see any reason why this lifecycle event should be used EXCEPT to delay the "Botpress is listening at" message. The studio should not be a dependency for anything, we will not ship it eventually.

Reply:
I just figured out, that when studio was ready, `process.NLU_PORT` was defined.

I can change it, but I believe I'll have to add an app lifecycle event ?

Reply:
Ok I added a `NLU_ENDPOINT_KNOWN` lifecycle event and it works fine